softmax/src/lib.rs

```diff
@@ -1,82 +1,191 @@
+
+        let input = TensorMetadata::new("input");
+        let hint = supported_shapes(
+            &[ElementType::F32, ElementType::F64],
+            Dimensions::Fixed(&[0]),
+        );
+        input.add_hint(&hint);
+        metadata.add_input(&input);
+
+        let soft_max = TensorMetadata::new("soft_max");
+        soft_max
+            .set_description("Vector normalised into probability distribution");
+        let hint = supported_shapes(
+            &[ElementType::F32, ElementType::F64],
+            Dimensions::Fixed(&[0]),
+        );
+        soft_max.add_hint(&hint);
+        metadata.add_output(&soft_max);
+
+        register_node(&metadata);
+    }
+
+    fn graph(id: String) -> Result<(), GraphError> {
+        let ctx = GraphContext::for_node(&id).ok_or_else(|| {
+            GraphError::Other("Unable to get the graph context".to_string())
+        })?;
 
-    fn transform(&mut self, input: Tensor<f32>) -> Tensor<f32> {
-        let b = input.map(|_, &x| expf(x as f32));
-        let sum: f32 = b.elements().iter().sum();
+        ctx.add_input_tensor("input", ElementType::F32, Dimensions::Dynamic);
 
-        b.map(|_, &x| x / sum)
+        ctx.add_output_tensor(
+            "soft_max",
+            ElementType::F32,
+            Dimensions::Dynamic,
+        );
+
+        Ok(())
     }
-}
 
-#[cfg(feature = "metadata")]
-pub mod metadata {
-    wit_bindgen_rust::import!(
-        "../wit-files/rune/runtime-v1.wit"
-    );
-    wit_bindgen_rust::export!(
-        "../wit-files/rune/rune-v1.wit"
-    );
-
-    struct RuneV1;
-
-    impl rune_v1::RuneV1 for RuneV1 {
-        fn start() {
-            use runtime_v1::*;
-
-            let metadata = Metadata::new("Softmax", env!("CARGO_PKG_VERSION"));
-            metadata.set_description("Apply [softmax](https://en.wikipedia.org/wiki/Softmax_function), sometimes referred to as the *\"normalized exponential function\"* to each element in the tensor.");
-            metadata.set_repository(env!("CARGO_PKG_REPOSITORY"));
-            metadata.set_homepage(env!("CARGO_PKG_HOMEPAGE"));
-
-            let input = TensorMetadata::new("input");
-            let hint =
-                supported_shapes(&[ElementType::Float32], Dimensions::Dynamic);
-            input.add_hint(&hint);
-            metadata.add_input(&input);
-
-            let output = TensorMetadata::new("output");
-            output.set_description("The result of `exp(x)` to each element, `x`, in the tensor and dividing by the sum.");
-            let hint =
-                supported_shapes(&[ElementType::Float32], Dimensions::Dynamic);
-            output.add_hint(&hint);
-            metadata.add_output(&output);
-
-            register_node(&metadata);
-        }
+    fn kernel(id: String) -> Result<(), KernelError> {
+        let ctx = KernelContext::for_node(&id).ok_or_else(|| {
+            KernelError::Other("Unable to get the kernel context".to_string())
+        })?;
```

Comment:
Same as `GraphError::MissingContext`.

```suggestion
        let ctx = KernelContext::for_node(&id).ok_or(KernelError::MissingContext)?;
```